### PR TITLE
pytorch plugin implementation

### PR DESCRIPTION
**Pull request by igorvalko**
# TL;DR
Pytorch plugin for Flyte that leverages https://github.com/kubeflow/pytorch-operator to run distributed pytorch jobs on k8s.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [X] Plugin

## Are all requirements met?

 - [X] Code completed
 - [ ] Smoke tested
 - [X] Unit tests added
 - [X] Code documentation added
 - [X] Any pending items have an associated Issue

## Complete description
Pytorch plugin for Flyte that leverages https://github.com/kubeflow/pytorch-operator to run distributed pytorch jobs on k8s.
Related PRs:
https://github.com/lyft/flyteidl/pull/61
https://github.com/lyft/flytekit/pull/112

## Tracking Issue
_NA_

## Follow-up issue
_NA_

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -11,7 +11,9 @@ require (
 	github.com/golang/protobuf v1.3.3
 	github.com/googleapis/gnostic v0.4.1 // indirect
 	github.com/hashicorp/golang-lru v0.5.4
-	github.com/lyft/flyteidl v0.17.9
+	github.com/kubeflow/pytorch-operator v0.6.0
+	github.com/kubeflow/tf-operator v0.5.3
```
Comment (kumare3) [inline go.mod:15]: do we need tf-operator?

File: go/tasks/plugins/k8s/pytorch/pytorch.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+	pluginsCore "github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/k8s"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/utils"
+
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flyteplugins/go/tasks/logs"
+	logUtils "github.com/lyft/flyteidl/clients/go/coreutils/logs"
+
+	//commonOp "github.com/kubeflow/common/pkg/apis/common/v1" // switch to real 'common' once https://github.com/kubeflow/pytorch-operator/issues/263 resolved
+	commonOp "github.com/kubeflow/tf-operator/pkg/apis/common/v1"
```
Comment (kumare3) [inline go/tasks/plugins/k8s/pytorch/pytorch.go:25]: aah i see that now!

File: go/tasks/plugins/k8s/pytorch/pytorch.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
+package pytorch
```
Comment (kumare3) [inline go/tasks/plugins/k8s/pytorch/pytorch.go:1]: can we move it a folder structure like
`k8s/kfoperators/pytorch/...` ?

File: go/tasks/plugins/k8s/pytorch/pytorch_test.go [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+		},
+	})
+	tID.On("GetGeneratedName").Return("some-acceptable-name")
+
+
+	resources := &mocks.TaskOverrides{}
+	resources.OnGetResources().Return(resourceRequirements)
+
+	taskExecutionMetadata := &mocks.TaskExecutionMetadata{}
+	taskExecutionMetadata.On("GetTaskExecutionID").Return(tID)
```
Comment (kumare3) [inline go/tasks/plugins/k8s/pytorch/pytorch_test.go:133]: we are trying to migrate everything to use
OnGetTraskExecutionID instead of On("GetTaskExecutionID") -> this helps in compile time safety. Can you please change all the "quoted" method references

**Comment by kumare3** [comment]
This is great stuff @igorvalko, once you change the flyteidl, you have a +1 from me

File: go/tasks/plugins/k8s/kfoperators/pytorch/pytorch.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "Unable to create pod spec: [%v]", err.Error())
+	}
+
+	overrideDefaultContainerName(taskCtx, podSpec)
+
+	workers := pytorchTaskExtraArgs.GetWorkers()
+
+	jobSpec := ptOp.PyTorchJobSpec{
+		TTLSecondsAfterFinished: nil, //60, // cleanup in 60 seconds after the job completion
```
Comment (kumare3) [inline go/tasks/plugins/k8s/kfoperators/pytorch/pytorch.go:77]: ya you dont need to cleanup, we will automatically

File: go/tasks/plugins/k8s/kfoperators/pytorch/pytorch.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+	pluginsCore "github.com/lyft/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/k8s"
+	"github.com/lyft/flyteplugins/go/tasks/pluginmachinery/utils"
+
+	"github.com/lyft/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/lyft/flyteplugins/go/tasks/logs"
+	logUtils "github.com/lyft/flyteidl/clients/go/coreutils/logs"
+
+	//commonOp "github.com/kubeflow/common/pkg/apis/common/v1" // switch to real 'common' once https://github.com/kubeflow/pytorch-operator/issues/263 resolved
```
Comment (kumare3) [inline go/tasks/plugins/k8s/kfoperators/pytorch/pytorch.go:24]: interesting call out. I think after this for all kubeflow operators, we can just have a `flyte-kubeflow-commons` where all status checking can be done inline
